src/crop-manager.sol

```diff
@@ -0,0 +1,164 @@
+    function frob(bytes32, address, address, address, int256, int256) external;
+    function flux(bytes32, address, address, uint256) external;
+    function hope(address) external;
+    function nope(address) external;
+}
+
+interface CropLike {
+    function gem() external view returns (address);
+    function ilk() external view returns (bytes32);
+    function join(address, address, uint256) external;
+    function exit(address, address, uint256) external;
+    function tack(address, address, uint256) external;
+    function flee(address) external;
+}
+
+interface TokenLike {
+    function approve(address, uint256) external;
+    function transferFrom(address, address, uint256) external;
+}
+
+contract UrnProxy {
+    address immutable public usr;
+
+    constructor(address vat_, address usr_) public {
+        usr = usr_;
+        VatLike(vat_).hope(msg.sender);
+    }
+}
+
+contract CropJoinManager {
+    mapping (address => uint256) public wards;
+    mapping (address => address) public proxy;  // UrnProxy per user
+    address public implementation;
+
+    event Rely(address indexed);
+    event Deny(address indexed);
+    event SetImplementation(address indexed);
+
+    constructor() public {
+        wards[msg.sender] = 1;
+        emit Rely(msg.sender);
+    }
+
+    function rely(address usr) public auth { wards[usr] = 1; emit Rely(msg.sender); }
+    function deny(address usr) public auth { wards[usr] = 0; emit Deny(msg.sender); }
+    modifier auth { require(wards[msg.sender] == 1, "CropJoinManager/non-authed"); _; }
+
+    function setImplementation(address implementation_) external auth {
+        implementation = implementation_;
+        emit SetImplementation(implementation_);
+    }
+
+    fallback() external {
+        address _impl = implementation;
+        require(_impl != address(0));
+
+        assembly {
+            let ptr := mload(0x40)
+            calldatacopy(ptr, 0, calldatasize())
+            let result := delegatecall(gas(), _impl, ptr, calldatasize(), 0, 0)
+            let size := returndatasize()
+            returndatacopy(ptr, 0, size)
+
+            switch result
+            case 0 { revert(ptr, size) }
+            default { return(ptr, size) }
+        }
+    }
+}
+
+contract CropJoinManagerImp {
+    bytes32 slot0;
+    mapping (address => address) proxy;  // UrnProxy per user
+
+    address public immutable vat;
+    constructor(address vat_) public {
+        vat = vat_;
+    }
+
+    function getOrCreateProxy(address usr) public returns (address urp) {
+        urp = proxy[usr];
+        if (urp == address(0)) {
+            urp = proxy[usr] = address(new UrnProxy(address(vat), usr));
+        }
+    }
+
+    function join(address crop, address urn, uint256 val) external {
+        TokenLike(CropLike(crop).gem()).transferFrom(msg.sender, address(this), val);
+        TokenLike(CropLike(crop).gem()).approve(crop, val);
+        CropLike(crop).join(getOrCreateProxy(urn), urn, val);
```

Comment:
Since `join` has `auth` now, we could in principle change it so that only a single `transferFrom` call is needed. Not something I feel strongly about but it would cut down on gas costs for users.

Reply:
The single `transferFrom` was what was done for the PSM. I've since changed my mind on that. I think it's cleaner / more intuitive to approve the manager. Additionally I believe there are gas changes coming that make zero -> non-zero -> zero storage updates really cheap which will be the case for the `approve` and provided there are no tokens accidentally transferred to the manager will be the case for 1 of the 2 `transferFrom`s as well.

Reply:
Sounds good.